스터디 API 작업

todayis-sunny opened the pull request:
## 📌 연결된 이슈
- #60 

### ✅ 작업 내용
#### 스터디 관련 API
- 스터디 **CRUD**
  - 스터디 생성 **C**
  - 스터디 목록 조회 **R**
  - 스터디 상세 조회 : **R**
  - 스터디 수정 **U**
  - 스터디 삭제 **D**
- 스터디 가입
  - 스터디 가입 신청
  - 스터디 가입 신청 취소
  - 스터디 멤버 가입 수락
  - 스터디 멤버 가입 거절
  - 스터디 멤버 목록 조회
- 스터디 멤버
  - 스터디 멤버 목록 조회
  - 스터디 멤버 권한 변경
  - 스터디 멤버 삭제

#### 카테고리 관련 API
- 카테고리 목록 조회
- 태그 목록 조회

#### 파일 API
- PreSigned Url 발급

### 📷 관련 이미지(영상)
| 스터디 생성 및 수정 | 스터디 목록 조회 | 스터디 상세 조회 |
| --- | --- | --- |
| <img width="300" alt="image" src="https://github.com/user-attachments/assets/8a9f0ada-0890-4052-9219-ac5e1854651b" /> | <img width="300" alt="image" src="https://github.com/user-attachments/assets/513d1591-ed7b-4d3c-9636-942ea8f7c3ef" /> | <img width="300" alt="image" src="https://github.com/user-attachments/assets/7d96e868-6567-4e47-a668-3cf92f087c71" /> | 

| 스터디 드로어 | 스터디 가입 | 스터디 멤버 관리 |
| --- | --- | --- |
| <img width="300" alt="image" src="https://github.com/user-attachments/assets/e2231b9e-425e-40fa-ae62-5b10e09963ca" /> | <img src="https://github.com/user-attachments/assets/bf84fb7a-aac9-462b-99a2-8330ec95d30e" width="300"> | <img width="300" alt="image" src="https://github.com/user-attachments/assets/588ad512-0383-493b-93b0-a33502c79566" /> |


### 🤔 의논하고 싶은 내용(코드 리뷰 받고 싶은 부분)
- 먼저 이슈단위를 크게 잡아버리고 작업속도도 느려서 스터디 API PR이 지체된점 송구스럽습니다😢
- 중간에 Detekt 반영된 `devlelop`를 머지하면서 제 작업이 아닌 부분도 **Files changed**에 포함되어 리뷰시 불편함을 겪으실 겁니다. 제가 Pakage를 보면서 제가 작업한 부분을 체크해두겠습니다!...
- **리팩토링 해야할 부분 많이 있습니다.** 먼저 API 테스트는 다 완료한 상태라 **PR**을 한번 진행하고 리팩토링을 수행하겠습니다.
---
- `MainScreen`단에서 드로어를 사용하면서 지저분해진 부분을 스크럼 시간때 피드백을 받았지만 최적화 하는 시간을 가져야할 것 같습니다.
- **스터디 가입신청**을 `StudiesCard`에서 할 수 있고, 내가 가입 신청한 스터디 목록을 알게 되면(서버와 협의하여 추가 구현 필요) 따로 구
분하여 제작하겠습니다. 현재는 **SnackBar**를 통해 이미 가입신청된 스터디 인 경우에 철회 가능합니다.
- 스터디 생성화면과 스터디 수정화면을 만들었고, 수정화면에만 Component를 일부분 쪼갰습니다. **Dropdown**을 component로 처리를 했는데, (Detekt에서 long method에 걸려서 분리하긴 했습니다...)
- 멤버 관리화면에서는 현재 직책 변경, 강퇴 두 가지 버튼이 존재하며 Dialog를 통해 관리 가능합니다.
- 가입 요청화면에서는 현재 수락, 거절 두 가지 버튼으로 바로 상호작용이 가능합니다.
- 그룹 삭제는 어디에 포함시켜야할지 의논해보면 좋을것 같습니다. 현재는 드로어에 테스트 용으로 두었으나, 그룹 관리 수정화면에 들어가는 기능이 될 수 있다고도 생각합니다.
- 스터디 목록 조회 화면에서 Paiging을 따로 사용하지 않았기 때문에, Paiging을 쓰신 분들을 보며 리팩토링 해보겠습니다.
- **stringResource**을 사용하는 부분도 있고 아닌 부분도 있는데, 의논해보고 정해지는 방향으로 리팩토링 해보겠습니다.

@upsk1, @Na2te, @Aloe-droid

todayis-sunny committed:
feat(#60): 카테고리, 태그 api 구현

todayis-sunny committed:
feat(#60): data, domain 레이어 스터디 관련 model 구현

Studies에서 유저에 의한 속성들은 StudyInfo로 분리

todayis-sunny committed:
feat(#60): 스터디 생성 api 구현

다른 api는 껍데기 구현

todayis-sunny committed:
refactor(#60): 스터디 스크린 디자인 시스템 적용

todayis-sunny committed:
fix(#60): 디자인 시스템 반영

라이트, 다크테마 디자인 시스템 미적용 부분 적용

todayis-sunny committed:
feat(#60): 스터디 수정 api 구현

ViewModel에는 미구현

todayis-sunny committed:
refactor(#60): selectedCategory withLoading에서 빼기

todayis-sunny committed:
feat(#60): 스터디 목록 조회 API 구현

커서 기반 페이징을 활용합니다.

todayis-sunny committed:
feat(#60): 스터디 상세 조회 API 구현

연관되어 있는 API 구현 필요(공지, 투표, 멤버, 피드)

todayis-sunny committed:
fix(#60): 스터디 관련 디자인 시스템 반영

다크, 라이트 테마에 맞게 적용

todayis-sunny committed:
feat(#60): file API 구현

PresignedUrl 발급 기능 구현

todayis-sunny committed:
feat(#60): S3 API 구현

PresignedUrl을 발급받아 이미지를 S3에 직접 업로드 하는 기능 구현

todayis-sunny committed:
fix(#60): studies API uri 구조 변경 수정

todayis-sunny committed:
chore(#60): file API, S3 API 구현 및 의존성 적용

todayis-sunny committed:
feat(#60): 스터디 프로필 이미지 등록 구현

PresignedUrl을 통한 프로필 이미지 업로드

todayis-sunny committed:
fix(#60): cursorCreatedAt 타입 String으로 변경

todayis-sunny committed:
fix(#60): studies API path variable 변경

카멜 case -> 케밥 case

todayis-sunny committed:
feat(#60): 스터디 가입 신청 API 구현

todayis-sunny committed:
chore(#60): dev 머지 후 변경 없이 detekt 검사 통과된 파일들

todayis-sunny committed:
refactor(#60): dev 머지 후 변경 detekt 검사 후 통과된 파일

- TimeFormatter renamed (클래스명과 멤버변수 네이밍이 동일)
- Device Locale 적용

todayis-sunny committed:
refactor(#60): dev 머지 후 변경 detekt 검사 후 통과된 파일

- Unit 제거 및 map 제거

todayis-sunny committed:
refactor(#60): dev 머지 후 변경 detekt 검사 후 통과된 파일

- 사용되지 않은 param 임시 사용 처리 (groupId, memberId, color)

todayis-sunny committed:
refactor(#60): dev 머지 후 변경 detekt 검사 후 통과된 파일

- Screen의 modifier을 Column에 적용
- 사용하지 않은 context 제거

todayis-sunny committed:
feat(#60): UpdateStudiesRequest 구현

todayis-sunny committed:
feat(#60): 스터디 신청 철회 API 구현

에러핸들링을 활용하여 가입된 스터디 or 가입 신청한 스터디인 경우 SnackBar를 통해 철회 -> 변경 필요

todayis-sunny committed:
feat(#60): 스터디 드로어 viewModel 공유 및 state 제어

- NavBackStackEntry를 활용하여 viewModel 공유

todayis-sunny committed:
feat(#60): 스터디 삭제 API 구현

- 스터디를 삭제하는 API 구현

todayis-sunny committed:
feat(#60): StudiesDrawer에서 스터디 삭제 기능 첨가

- 스터디 삭제를 드로어에서 젹용 (변경 필요)

todayis-sunny committed:
feat(#60): 스터디 삭제 Dialog 구현

- 스터디 삭제 요청을 확인하는 Dialog 구현

todayis-sunny committed:
feat(#60): 스터디 멤버 목록 조회 API 구현

- API, DataSource, Repository, UseCase 구현
- Mapper를 적용하여 데이터 변환

todayis-sunny committed:
feat(#60): 스터디 멤버 목록 조회 API 사용

- 기본적인 userId, name, profileImg, groupRole을 받아옴
- 추후 목표 공부시간에 따른 학습시간 원형 링(비율) 데이터 필요

todayis-sunny committed:
feat(#60): 스터디 가입 대기 인원 목록 조회 API 구현

- API, DataSource, Repository, UseCase 구현
- data, domain에 모델 선언

todayis-sunny committed:
feat(#60): 스터디 가입신청 수락 API, 스터디 가입신청 거절 API 구현

- API, DataSource, Repository, UseCase 구현
- notificationId는 null처리 (알림이 아닌 화면 이동으로 처리하기 때문)

todayis-sunny committed:
feat(#60): 스터디 가입신청 목록조회, 수락, 거절, 로직 구현

- cursor기반 가입신청 목록 조회
- patch방식의 가입신청 수락, 가입신청 거절 구현
- 가입신청 아이템 component 구현
- enum class를 활용하여 상태 구분(단발성 조회 -> 사용자의 행동에 따른 페이지 변화[대기, 수락, 거절])

todayis-sunny committed:
feat(#60): 가입신청 현황 드로어에 연결

- NavController를 직접 MainScreen에서 주입(NavGraph보다 상단에 드로어가 있기 때문)
- 드로어 item 추가

todayis-sunny committed:
feat(#60): 스터디 멤버 권한 변경 API, 강퇴 API 구현

- API, DataSource, Repository, UseCase 구현

todayis-sunny committed:
feat(#60): 스터디 멤버 관리 기능 구현(권한 변경, 강퇴)

- Contract, Screen, ViewModel 구현
- 단위 component 구현

todayis-sunny committed:
feat(#60): 스터디 멤버 권한 변경 및 강퇴 Dialog 구현

todayis-sunny committed:
refactor(#60): StudiesMember domain model 역할 타입 수정

- String -> enum class

todayis-sunny committed:
feat(#60): preview 용도 getStudiesMembers 구현

todayis-sunny committed:
feat(#60): 멤버 관리 드로어에 적용

Na2te committed:
feat(#86): Color 수정
- 기존 Color 파일의 neegongnaegongcolors로 되어 있던 것이 detekt에 막혀 Color로 수정
- 투표시 버튼 활성화 색깔인 yellow 추가

todayis-sunny committed:
refactor(#60): 스터디 생성 UI수정 및 네이밍 변경

- Management -> Create로 변경(Edit은 따로 있기 때문)
- 생성 유효성 검사 로직 추가
- 최대 입력 글자수 추가

todayis-sunny committed:
feat(#60): 스터디 수정 기능 구현

- component 분리 (드롭다운, 이미지피커)

todayis-sunny committed:
fix(#60): 스터디 목록 조회 API query param 네이밍 변경

- (cursorId)camel case -> (cursor-id)snake case

todayis-sunny commented:
아직 리뷰해주시지 마시고
컨플릭트도 해결해두고 File Changed 패키지 구분해두겠습니다!

upsk1 commented:
![image](https://github.com/user-attachments/assets/4b6bb56d-c787-484f-8e3d-97aec5d4c8ce)

저 아직 소스보기 전인데 진짜 대충돌사고가 났네요 ㅋㅋㅋ

todayis-sunny committed:
chore(#60): Detekt 검사 후 일부 파일 수정

- datOfWeekOrder.kt(앞에 대문자 처리)
- BaseViewModel.kt(event 접근제한자, handleException 네이밍)
- NotificationScreen.kt(image string 주석 처리)
- ProfileText.kt (const val SNAKE CASE)
- PreviewAnnotation.kt (파일 네이밍 변경)

todayis-sunny committed:
chore(#60): pre-commit 주석해제

Na2te commented:
아직 파일을 보진 않았는데 코드 양만 봐도 얼마나 고생하신 건지 감히 평가할 수 없는 양이네요

작업 속도는 사람마다 기준이 다를 거라고 생각합니다
PR이 작업 시작 시점 기준으로 빠르진 않았다고 할 수 있지만,
말씀하신대로 이슈 크기가 워낙 방대한 양을 다 포함하고 있었고, 개발 양을 고려했을 때 이걸 누가 늦었다고 할 수 있을까요
심지어 detekt 및 conflict 관련해서 해결해야 하는 본인 개발 외의 문제들도 있었던 거로 알고 있습니다

홀로 성실, 묵묵히 작업하시는 걸 신경 쓰지 못했던 업보가 팀이 아니라 되려 본인의 귀책처럼 느껴지게 만든 것 같아 죄송하네요
그러니 미안한 마음을 가지시진 않으셨으면 좋겠습니다

파일 change 184개는 무섭긴 한데, 업보라고 생각하고 달..게 받아보겠..습니다... :cold_sweat:

추신 - 혹 컨플릭트 관련 도움 필요하시면 말씀 주세용

todayis-sunny committed:
fix(#60): LearningRecordApi Hilt 어노테이션 누락 수정

todayis-sunny commented:
<img width="1666" alt="image" src="https://github.com/user-attachments/assets/bf07241f-05a0-481e-a3d0-6ad31306bc99" />
해당 이미지 내역만 Detekt 수정했습니다!...

todayis-sunny commented:
컨플릭트 해결은 하고 잠시 쉬고 왔습니다...
@upsk1 
아직은 읽기에 불편하실겁니다!

@Na2te 
말씀 또한 감동있게 하십니다...😊
그래도 이번 사태로 인해서 작업을 작게 잡고 스크럼에서 말했다시피 PR의 양도 고려하면 해야겠다고 느낍니다.

방금 해결을 하고 오긴 했는데 아직 컨플릭트 나는게 #117 이 머지되어서 그런거겠죠?
제가 작업한 패키지 한번 더 해결하고 패키지 정리해서 올려보겠습니다!

todayis-sunny committed:
chore(#60): conflict 해결, Detekt 검사

- NoticeCard.kt (modifier param 사용)

Na2te commented:
> 방금 해결을 하고 오긴 했는데 아직 컨플릭트 나는게 #117 이 머지되어서 그런거겠죠? 제가 작업한 패키지 한번 더 해결하고 패키지 정리해서 올려보겠습니다!

해결하셨는데 그런거면 그럴 것 같습니다..
머지를 빠르게 할 걸 그랬네용

todayis-sunny commented:
@Na2te
아닙니다! 잘 해결했습니다! (아마도...)

todayis-sunny commented:
<img width="964" alt="image" src="https://github.com/user-attachments/assets/2a8b53f3-9e48-4deb-8579-dffc900c87ee" />
- 해당 커밋 내역에서 **Detekt**  키워드가 들어가 부분 빼고 봐주시면 나름 보기 편하실 수 있습니다!...
  - 해당 부분을 드래그해서 커밋을 범위지정해서 볼수 있습니다.
 
<img width="954" alt="image" src="https://github.com/user-attachments/assets/15919427-4887-4d59-a478-67090b3c9f7c" />
- 패키지 검색 부분에 `Studies`를 검색해서 보시는것도 좋을것 같습니다.
- 대부분 `Studies`이 들어간 패키지에 많이 있습니다.
- 추가로 `S3`, `Category`, `Tag` 봐주시면 좋을것 같습니다.

@upsk1, @Na2te, @Aloe-droid 
혼란스럽게 해서 미안할 따름입니다...😢

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/data/model/studies/request/CreateStudiesRequest.kt:
```diff
@@ -0,0 +1,28 @@
+
+import com.ssafy.neegongnaegong.domain.model.studies.StudyInfo
+
+data class CreateStudiesRequest(
+    val name: String,
+    val maxMembers: Int,
+    val description: String,
+    val profileImg: String?,
+    val isPublic: Boolean,
+    val targetStudyTime: Int,
+    val category: Long,
+    val tags: List<Long>,
+) {
+    companion object {
+        fun fromDomain(studyInfo: StudyInfo) =
+            CreateStudiesRequest(
+                name = studyInfo.name,
+                maxMembers = studyInfo.maxMembers,
+                description = studyInfo.description,
+                profileImg = studyInfo.profileImg,
+                isPublic = studyInfo.isPublic,
+                targetStudyTime = studyInfo.targetStudyTime,
+                category = studyInfo.category?.id ?: 0,
+                tags = studyInfo.tags.map { it.id },
+            )
+    }
+}
```
저희가 Mapper 클래스로 정의해서 Domain <-> Data 데이터를 변환하기로 결정하였습니당..
현재는 문제가 되지 않으니 추후 리팩토링 때 한번에 Mapper 클래스를 만들면 좋을 것 같습니다!

Na2te replied:
@Aloe-droid 
이거 관련해서 조금 여쭙고 싶은게 있었는데, 현재 변환 함수들이 전부 클래스의 확장함수여서 Mapper Object 자체가 쓰이지가 않는 것 같습니다
그래서 Mapper Object의 의의가 없는 것 같고, 그냥 data class 파일 하단부에 확장함수를 쓰는게 더 나아보이는 상황 같은데 이 부분에 대해서 생각이 궁금합니다

upsk1 replied:
저도 이 부분 약간 어색하게 생각했었습니다 , Object면 항상 존재할텐데 이러면 메모리누수가 있지 않을까 싶었습니다 
영민님 말대로 하단부 확장함수로 사용할지 class로 사용할지에 대한 고민도 좋아보입니다

Aloe-droid replied:
좋은 의견이네여
`Object` vs `Companion Object` 모두 싱글톤 형태이지만 초기 적재 시기만 다른 것으로 알고 있습니다. 따라서 유의미한 성능 차이가 있을지는.. 없다고 생각합니다. 
특히 `Mapper` 클래스의 특성상 변수는 없고 함수로만 존재할텐데 적재된 메모리의 양도 크지 않을 것이라고 생각합니다. 

`Mapper` 클래스를 두는 이유는.. Domain <-> Data 모델간 변환이 복잡해질 수록 `Data Layer`의 데이터 클래스가 불필요가 커지기 때문이라고 생각합니다. 
예를들어 `Data(A) -> Domain(A)` 뿐 아니라 `Data(List<A>) -> Domain(List<A>)` , `Data(PagingData<A>) -> Domain(PagingData<A>)` 와 같은 매핑 함수를 정의하게 되면 `Data(A)` 클래스의 하단에 여러 함수들을 정의해야 합니다. 
이 경우 말씀해주신 부분처럼 A 클래스 안에 매핑을 정의해도 괜찮지만, 관심사를 분리해서 `(데이터) , (매퍼)` 를 분리하는 것도 좋은 방식이라고 생각합니다. 

결론적으로는 성능상의 이슈는 차이가 없을 것이라고 생각하고, 관심사를 분리하는 관점에서는 매퍼 분리가 좋다고 생각합니다.

todayis-sunny replied:
리팩토링 이슈로 만들어 **Mapper**로 변환과정 갖추겠습니다!

Na2te replied:
그렇다면 Mapper 내부에 확장 함수가 아니라 (Data layer model) -> (Domain layer model) 꼴의 Data Layer 모델의 인자를 받아서 Domain Layer 모델로 변환하는 함수로 만들어서, Mapper Object가 관련 책임을 명확하게 가지도록 해야 할 것 같은데 어떻게 생각하시나요??

todayis-sunny replied:
>그렇다면 Mapper 내부에 확장 함수가 아니라 (Data layer model) -> (Domain layer model) 꼴의 Data Layer 모델의 인자를 받아서 Domain Layer 모델로 변환하는 함수로 만들어서, Mapper Object가 관련 책임을 명확하게 가지도록 해야 할 것 같은데 어떻게 생각하시나요??

혹시 제가 답변드리는게 맞을까요?

Na2te replied:
아 딱히 누구를 지칭하고 물어본건 아닙니다!
Aloe님 의견에 덧붙인거긴 한데 다들 의견이 있으면 말하면 되죵

todayis-sunny replied:
아하, 확장함수로 하게되면 Domain model에 책임이 있는 형태이므로
```kotlin
fun toDmain(argm: Data): Dmoain {
// 구현
}
```
이런 구조로 가져가서 Mapper가 책임을 진다는 의미인걸까요?
저는 좋은 내용인거 같습니다. 동의힙니다!

todayis-sunny replied:
추가로 Domain model -> Data layer인 경우에는 확장함수를 쓰는걸까요?
아니면 둘다 일반함수로 하는 의미일까요?

Aloe-droid replied:
> 그렇다면 Mapper 내부에 확장 함수가 아니라 (Data layer model) -> (Domain layer model) 꼴의 Data Layer 모델의 인자를 받아서 Domain Layer 모델로 변환하는 함수로 만들어서, Mapper Object가 관련 책임을 명확하게 가지도록 해야 할 것 같은데 어떻게 생각하시나요??

Mapper 안에 확장함수로 정의하냐 일반 함수로 정의하냐에 대한 질문이시라면 큰 차이를 잘 못느끼겠네요..
둘을 비교했을때 유의미한 장단점도 떠오르지 않아서... 편하신대로 만들면 되지 않을까 싶습니다.

Na2te replied:
> 아하, 확장함수로 하게되면 Domain model에 책임이 있는 형태이므로
> 이런 구조로 가져가서 Mapper가 책임을 진다는 의미인걸까요? 저는 좋은 내용인거 같습니다. 동의힙니다!

아 아니용 확장함수는 멤버 변수가 아닙니당
그래서 확장함수 자체는 해당 클래스에 대해 아무런 책임을 가지지 않습니당

그래서 확장함수 말고 일반 함수로 해야 하지 않을까 와 같은 얘기였습니당

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/domain/model/preview/studies/StudiesPreviewDataProvider.kt:
```diff
@@ -0,0 +1,142 @@
+import com.ssafy.neegongnaegong.presentation.group.role.component.StudiesMemberRole
+import com.ssafy.neegongnaegong.presentation.util.TimeUnit
+
+class StudiesPreviewDataProvider {
+    fun getStudies(): List<Studies> =
```
Preview 데이터를 빼놓을 수도 있군여 ㄷㄷ 배웠습니다

todayis-sunny replied:
@upsk1 의견으로 빼두긴 했는데 다 빼두진 못 했습니다🫨

jaehan4707 replied:
https://developer.android.com/develop/ui/compose/tooling/previews?hl=ko

다음과 같이 previewParameter 정의할 수도 있으니 참고하면 좋을것 같습니다 ㅎ

todayis-sunny replied:
`previewParameter`가 있군요! 배워갑니다 ㅎㅎ

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/module/di/NetworkModule.kt:
```diff
@@ -51,84 +54,148 @@ object NetworkModule {
+    ): Retrofit =
+        Retrofit
+            .Builder()
+            .baseUrl("https://dummy/")
```
s3 주소의 시작이 dummy 인가요 신기하네요

todayis-sunny replied:
이 부분은 제가 작성하면서 Retrofit을 빌더 패턴으로 만들어 둘때 baseUrl을 넣어야하는 의무가 있었고 의미없는 데이터를 넣어두고
```kotlin
interface S3Api {
    @PUT
    suspend fun uploadImageToS3(
        @Url presignedUrl: String,
        @Body body: RequestBody,
    ): Response<Unit>
}
```
API에서 `@Url`가 덮어 쓰는 형태로 진행되었습니다. 먼가 저 baseUrl을 부분을 깔끔하게 적고 싶긴하네요...

Aloe-droid replied:
아 retrofit 의 빌더 패턴 중 baseUrl이 필수 파라미터인가요? 없이 작성했을 때가 궁금하네여

todayis-sunny replied:
궁금하신점 일단 보여드리겠습니다!
- 주석처리 해보겠습니다.
![image](https://github.com/user-attachments/assets/84b87d2e-bed8-43c5-ae01-92f3ae0f31f4)
- 로그입니다.
![image](https://github.com/user-attachments/assets/1f6a3cd1-1ae4-4cc2-bcf8-76e841c0c2b2)
- AI한테 얻은 자문입니다.
<img width="653" alt="image" src="https://github.com/user-attachments/assets/cec2248e-19f2-40cc-9be4-d8722632b6a5" />

앱 실행시에는 괜찮았고, S3API를 사용하는 스터디 사진이 등록가능환 하면에 들어갈 때 **runtime error**가 발생했습니다.
`baseUrl()` 기본적으로 있어야 하나 봅니다 ㅠ

Aloe-droid replied:
그렇군요... 배웠습니다 감사합니다

Na2te replied:
오 S3 URL이 공통된 Prefix부분이 아예 없나보네용 저도 배워갑니당

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/MainScreen.kt:
```diff
@@ -131,16 +141,39 @@ fun MainScreen() {
-                onMyStudyClick = {},
-                onStudyItemClick = {},
-            )
+            navBackStackEntry?.let { entry ->
+                val isStudiesDetail =
+                    entry.destination.hierarchy.any {
+                        it.hasRoute(AppNavigation.Screen.Studies.StudiesDetail::class)
+                    }
+                StudiesDrawerContent(
+                    navBackStackEntry = entry,
+                    navigateTodStudiesEdit = {
+                        if (isStudiesDetail) {
+                            val route = entry.toRoute<AppNavigation.Screen.Studies.StudiesDetail>()
+                            navController.navigate(
+                                AppNavigation.Screen.Studies.Edit(route.studyGroupId),
+                            )
+                        }
+                    },
+                    navigateToStudiesMembersRole = {
+                        if (isStudiesDetail) {
+                            val route = entry.toRoute<AppNavigation.Screen.Studies.StudiesDetail>()
+                            navController.navigate(
+                                AppNavigation.Screen.Studies.StudiesMembersRole(route.studyGroupId),
+                            )
+                        }
+                    },
+                    navigateToStudiesApplications = {
+                        if (isStudiesDetail) {
+                            val route = entry.toRoute<AppNavigation.Screen.Studies.StudiesDetail>()
+                            navController.navigate(
+                                AppNavigation.Screen.Studies.StudiesApplication(route.studyGroupId),
+                            )
+                        }
+                    },
+                )
+            }
```
만약 isStudiesDetail 가 true일 때에만 화면에 drawer 가 보여지고 싶으면 `return` 키워드를 사용하면 block 하나를 줄일 수 있을 것 같습니다. 
```kotlin
drawerContent = {
    val isStudiesDetail: Boolean = navBackStackEntry
        ?.destination
        ?.hierarchy
        ?.any { navDestination: NavDestination ->
            navDestination.hasRoute(AppNavigation.Screen.Studies.StudiesDetail::class)
        } == true
    
    if(!isStudiesDetail) return@ModalNavigationDrawer
    StudiesDrawer(
        headerImageUrl = null,
        onGroupManagementClick = {},
        onMemberManagementClick = {},
        onScheduleManagementClick = {},
        onStudyCreateClick = {},
        onStudySearchClick = {},
        onMyStudyClick = {},
        onStudyItemClick = {},
    )
},
```

todayis-sunny replied:
좋은 의견 감사합니다. 반영 해보겠습니다.

todayis-sunny replied:
해당 부분으로 처리해보니 깜빡이는 현상 때문에 기존 코드로 진행하고 나중에 심화 리팩토링 진행해보겠습니다! 🥲

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/base/BaseViewModel.kt:
```diff
@@ -37,7 +37,7 @@ abstract class BaseViewModel<Event : UiEvent, State : UiState, Effect : UiEffect
 
     private val _event: MutableSharedFlow<Event> = MutableSharedFlow()
-    private val event = _event.asSharedFlow()
+    val event = _event.asSharedFlow()
```
기존에는 `event`를 `private`으로 사용하고 `handleEvent`로 정의했는데, 접근제한자가 변경된 이유가 있나여??

todayis-sunny replied:
https://github.com/NeeGongNaeGong/android/pull/121#discussion_r2141715758
여기 코멘트 확인 한번 부탁드립니다.
Detekt에서 막아버리던데 혹시 더 나은 방법이 있을까요?🧐

Aloe-droid replied:
Detekt를 잘 몰라서 조언이 어렵네요... ㅠㅠ

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/base/BaseViewModel.kt:
```diff
@@ -244,34 +246,38 @@ abstract class BaseViewModel<Event : UiEvent, State : UiState, Effect : UiEffect
-        retry: suspend () -> Unit
+        retry: suspend () -> Unit,
     ) {
         when (e) {
```
`_handleException` 는 `BaseViewModel` 에서만 접근하는 함수로 공통 에러(액세스 토큰..)를 처리하고,  `handleException` 으로 개별적 `viewModel`에서 정의하여 구현했는데, 이렇게 함수명이 하나로 되어버리면 문제가 생길 것 같은데 확인해주시면 감사하겠습니당

todayis-sunny replied:
https://github.com/NeeGongNaeGong/android/pull/121#discussion_r2141718461
Detekt에서 검사 걸려 수정을 했는데
제가 수정한 사항에도 문제를 인지했습니다. 네이밍 수정이 필요할 것 같습니다.

Na2te reviewed (commented):
ef3585afbe26ef2d87865a3829fb861b45bece4c 이 커밋을 봤는데, 수정사항이 다 제 것 관련인 거 같은데 문제 없어보입니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/base/BaseViewModel.kt:
```diff
@@ -37,7 +37,7 @@ abstract class BaseViewModel<Event : UiEvent, State : UiState, Effect : UiEffect
 
     private val _event: MutableSharedFlow<Event> = MutableSharedFlow()
-    private val event = _event.asSharedFlow()
+    val event = _event.asSharedFlow()
```
이벤트가 private으로 public으로 바뀐 이유가 있나요??

todayis-sunny replied:
아, 이부분도 Detekt검사에서 보더라구여 
![image](https://github.com/user-attachments/assets/bc45ef1a-c4f3-4db7-9995-c71832a5aa72)
2번째 줄 **BaseViewModel.kt 39행** 입니다.
<img width="500" alt="image" src="https://github.com/user-attachments/assets/fe0269e7-007a-439e-b618-10659bce4615" />

우선 위에 변수들의 접근 제한자처럼 통일 해둿습니다!
다른 방법이 있을까요 혹시?

Na2te replied:
아 음 확인했습니다
외부에 공개 안하니까 그냥 `event` 객체를 삭제하고 `ViewModel` 내부에서 `collect` 및 `emit`을 _event 하나로만 하는게 좋지 않을까 싶네요
`_event`라고 하면 naming 뭐라고 할 테니 아마 `event`라고 이름도 변경해야 될 것 같기도 하는데 이건 다른 사람들 의견도 들어봐야 할 것 같습니다

todayis-sunny replied:
넵, 확인했습니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/base/BaseViewModel.kt:
```diff
@@ -244,34 +246,38 @@ abstract class BaseViewModel<Event : UiEvent, State : UiState, Effect : UiEffect
         errorContext: ErrorContext?,
-        retry: suspend () -> Unit
+        retry: suspend () -> Unit,
     ) {
```
이 부분은 ViewModel에서 override 해서 쓰는 handelException하고 혼동의 소지가 있으므로, defaultHandleException 이런 거로 renaming 하는게 좋을 거 같습니다!

todayis-sunny replied:
그렇네요 `_`를 괜히 넣은게 아닌데, 네이밍 수정 하겠습니다!

Aloe-droid reviewed (commented):
양도 정말 방대하고 viewModel의 로직들을 보면서 얼마나 힘들어하셨을지 감도 안잡히네요 
정말 고생하셨습니다 👏

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/data/datasource/network/NetworkStudiesDataSource.kt:
```diff
@@ -14,4 +22,48 @@ interface NetworkStudiesDataSource {
         requestBody: CreateNoticeRequest,
     ): Flow<Unit>
+
+    suspend fun getStudiesList(request: GetStudiesListRequest): Flow<CursorSliceStudiesListResponse>
+
+    suspend fun createStudies(request: CreateStudiesRequest): Flow<Unit>
+
+    suspend fun getStudies(studyGroupId: Long): Flow<StudiesResponse>
```
혹시 여유되실 때 suspend 랑 Flow랑 둘중 하나만 써주실 수 있을까요 ?!

todayis-sunny replied:
**suspend**랑 **Flow** 둘 중 하나만 사용하겠습니다!

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/data/model/studies/request/CreateStudiesRequest.kt:
```diff
@@ -0,0 +1,28 @@
+
+import com.ssafy.neegongnaegong.domain.model.studies.StudyInfo
+
+data class CreateStudiesRequest(
+    val name: String,
+    val maxMembers: Int,
+    val description: String,
+    val profileImg: String?,
+    val isPublic: Boolean,
+    val targetStudyTime: Int,
+    val category: Long,
+    val tags: List<Long>,
+) {
+    companion object {
+        fun fromDomain(studyInfo: StudyInfo) =
+            CreateStudiesRequest(
+                name = studyInfo.name,
+                maxMembers = studyInfo.maxMembers,
+                description = studyInfo.description,
+                profileImg = studyInfo.profileImg,
+                isPublic = studyInfo.isPublic,
+                targetStudyTime = studyInfo.targetStudyTime,
+                category = studyInfo.category?.id ?: 0,
+                tags = studyInfo.tags.map { it.id },
+            )
+    }
+}
```
타입변환은 mapper에서 하시는게 어떨까요 ?!

todayis-sunny replied:
넵 Mapper로 변환하겠습니다!

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/data/model/studies/response/CursorSliceStudiesListResponse.kt:
```diff
@@ -0,0 +1,18 @@
+    val cursorCreatedAt: String,
+    val cursorId: Long,
+) {
+    fun toDomain(): CursorStudiesPage =
+        CursorStudiesPage(
+            content = content.map { it.toDomain() },
+            hasNext = hasNext,
+            cursorCreatedAt = cursorCreatedAt,
+            cursorId = cursorId,
+        )
```
Domain -> Data 
Data -> Domain 
둘 다 하나의 데이터클래스에서 변환책임을 주지말고 mapper에서 주기로 하자 라는 의견이 있어서 이 부분도 바꿀만 하지 않나요 ?!

todayis-sunny replied:
네 이부분도 수정하겠습니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/notification/NotificationScreen.kt:
```diff
@@ -103,28 +107,28 @@ fun NotificationScreen(
-        studyGroupId = null,
-        studyGroupName = null
-    )
+    // image = "https://encrypted-tbn1.gstatic.com/images?q=tbn:ANd9GcTRI8A6M23RTePWn8of5fgwRzSMMzRy_6mZP7OrP79VF3ByzCoRcyfx6bYr9w4bH9zdVfpV_LP9hBAudM5SRGyjnbbEhnrs2vWZKF8wySI",
```
이미지 안쓰는 거면 삭제해도 될 것 같습니당
근데 이미지 테스트 용으로 있었던 거 같은데, 안쓰는 이유가 있나용?

todayis-sunny replied:
여기가 @Aloe-droid 작업이라 지우기 좀 그래서 주석으로 해두긴 했습니다.🫠
저번에 개인으로 연락드려서 Detekt 검사를 잠시 off 할 수 없는지 여쭙기 전에 제가 일부분 수중한건데 Detekt에서 내용이 너무 길다고 해서 처리해두었습니다!

Na2te replied:
string값이라서 
경로  = "일부분"
+"일부분"

요런식으로 하면 될 거 같습니다!

todayis-sunny replied:
시도해보겠습니다!

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/domain/model/preview/studies/StudiesPreviewDataProvider.kt:
개인적으로 이부분 되게 좋아보입니다 ㅎㅎ

todayis-sunny replied:
처음부터 저렇게 했으면 좋았을것 같은데, 모든 곳에 적용하지는 못 했습니다ㅋㅋ 🫨

Na2te reviewed (commented):
e02df67d78b20df59a868a2bfe8aa4548b3f4b6e 커밋에 수정사항을 알려주신 덕분에 진짜 빠르게 필요한 부분만 확인했습니당

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/component/drawer/StudiesDrawer.kt:
```diff
@@ -25,42 +26,73 @@ import androidx.compose.ui.Modifier
+) {
+    val viewModel: StudiesDetailViewModel = hiltViewModel(navBackStackEntry)
+    val uiState = viewModel.uiState.collectAsStateWithLifecycle()
+    Log.d(TAG, "img : ${uiState.value.studies.studyInfo.profileImg}")
```
이 부분도 여유가 되실떄 슬쩍 지워주실 수 있나요 ?!

todayis-sunny replied:
넵, 알겠습니다👌

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/create/StudiesCreateContract.kt:
```diff
@@ -0,0 +1,87 @@
+import com.ssafy.neegongnaegong.presentation.base.UiState
+import okhttp3.RequestBody
+
+class StudiesCreateContract {
+    sealed class Event : UiEvent {
+        data object OnLoad : Event()
+
+        data class OnNameChanged(
+            val name: String,
+        ) : Event()
+
+        data class OnIsPublicChanged(
+            val isPublic: Boolean,
+        ) : Event()
+
+        data class OnTargetStudyTimeChanged(
+            val targetStudyTime: Int,
+        ) : Event()
```
https://github.com/NeeGongNaeGong/android/pull/119#discussion_r2138077103 

영민님께서 알려주신 부분인데, 상속할때 공통적으로 넣어줄 파라미터가 없으면 sealed interface로 해서 코드를 더 깔끔히 하는게 있습니다 !

Na2te replied:
맞습니다 저렇게 할 수도 있습니다
다만 https://github.com/NeeGongNaeGong/android/pull/118#discussion_r2138036781 이러한 의견도 있기에 꼭 해야 한다기보다 일단은 취향껏 하셔도 될 것 같습니다

todayis-sunny replied:
맞습니다 저도 예전에 코멘트로 물어본적이 있었는데, 새로 만들때는 `interface`를 활용했었는데,
리팩토링 이슈를 거쳐서
https://github.com/NeeGongNaeGong/android/pull/117#discussion_r2139122587
의 **import** 방식으로 아래 내용도 `interface`와 한번에 같이 바꾸려고 합니다!

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/create/StudiesCreateContract.kt:
```diff
@@ -0,0 +1,87 @@
+        val studyInfo: StudyInfo = StudyInfo.empty(),
+    ) : UiState
+
+    sealed class Effect : UiEffect {
+        data object NavigateToBack : Effect()
+    }
+
+    sealed class Error : ErrorContext {
+        data object CreateStudiesError : Error()
+    }
```
https://github.com/NeeGongNaeGong/android/pull/119#discussion_r2138077103

마찬가지로 이부분도 sealed interface로 좀 더 간결히 하실 수 있을 것 같습니다!

todayis-sunny replied:
넵, 수정하겠습니다!

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/create/StudiesCreateViewModel.kt:
```diff
@@ -0,0 +1,214 @@
+                issuePresignedUrlUseCase(
+                    uploadPathType = UploadPathType.PROFILE_STUDY_TEMP.path,
+                    imageExtension = extension,
+                ).withLoading {
+                }.safeCollect { result ->
```
withLoading 이 이렇게 써도 작동하나여 ?!

todayis-sunny replied:
네, 사실은 `isLoding` 변수 같은 곳에 **ture** -> **false**를 전달해주기 위함으로 사용된다고 알고 있습니다.
로딩 부분을 체크를 못 해었네요... 잘 작동합니다!

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/detail/StudiesDetailContract.kt:
```diff
@@ -0,0 +1,30 @@
+import com.ssafy.neegongnaegong.presentation.base.UiState
+
+class StudiesDetailContract {
+    sealed interface Event : UiEvent {
+        data class OnLoad(
+            val studyGroupId: Long,
+        ) : Event
+
+        data class OndDeleteStudies(
+            val studyGroupId: Long,
+        ) : Event
+    }
```
엇 이부분은 sealed interface가 잘 작성되있네요 ?!

todayis-sunny replied:
맞습니다. 앞서 남겼던 내용처럼 저도 새로 만드는 곳은 `interface`로 작성중에 있었습니다.

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/data/datasource/network/NetworkCategoryDataSource.kt:
```diff
@@ -0,0 +1,11 @@
+import com.ssafy.neegongnaegong.domain.model.studies.Tag
+import kotlinx.coroutines.flow.Flow
+
+interface NetworkCategoryDataSource {
+    suspend fun getCategories(): Flow<List<Category>>
+
+    suspend fun getTags(categoryId: Long): Flow<List<Tag>>
+}
```
요 부분은 https://github.com/NeeGongNaeGong/android/pull/121#discussion_r2141622692 이슈랑 동일합니다
suspend를 빼도 무방할 것 같습니다!

todayis-sunny replied:
넵, **suspend**를 제거하겠습니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/data/remote/CategoryApi.kt:
```diff
@@ -0,0 +1,17 @@
+import retrofit2.http.Path
+
+interface CategoryApi {
+    @GET("/api/categories")
```
제가 StudyGroup쪽 관련해서 만들었는데, /api/categories/ 라는 접두사가 공통된다면 Prefix로 빼는 것도 한번 생각해보실만 한 것 같습니다

todayis-sunny replied:
https://github.com/NeeGongNaeGong/android/pull/121#discussion_r2141708881
여기왜 Reply이 안뜨죠?... 여기에 남깁니다.
이 부분도 Mapper로 변경해야하는걸로 동의합니다. 수정하겠습니다!

todayis-sunny replied:
> 제가 StudyGroup쪽 관련해서 만들었는데, /api/categories/ 라는 접두사가 공통된다면 Prefix로 빼는 것도 한번 생각해보실만 한 것 같습니다

맞습니다 저도 PREFIX 방식으로 하려고 보고 같이 하려고 했습니다.
<img width="722" alt="image" src="https://github.com/user-attachments/assets/4704a492-94fd-4649-b7a9-a6209262d302" />
단순하게 `interface class`로 인지 못하는 점에서 사용하지 않았었는데요,

PREFIX를 다시 어떻게 활용할 수 있을까 하다가
API 내부에 `companion object` 안에 상수를 투입하는건 어떨까 싶습니다.
<img width="793" alt="image" src="https://github.com/user-attachments/assets/0030eacf-7a53-4be5-921f-7240c79b82b9" />
이러면 `interface class`로 인지를 잘 할수 있고, 다른 외부 파일에서 PREFIX를 혼동하지 않고 사용할 수 있을것 같습니다.
현재는 **StudyGroupAPI**의 `const val PREFIX = "/api/study-groups"`이 다른 외부 파일에서도 참조가 가능한 상태입니다.

아니면 공통된 접두사를 사전에 다른곳에 정의할수 있는곳이라면 좋을거 같습니다.

Na2te replied:
아이고 제가 잘못했네용 companion이 맞을 것 같습니다!

todayis-sunny replied:
`companion object`로 작성해보겠습니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/data/repository/CategoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,28 @@
+        private val dataSource: NetworkCategoryDataSource,
+        @IoDispatcher private val ioDispatcher: CoroutineDispatcher,
+    ) : CategoryRepository {
+        override suspend fun getCategories(): Flow<List<Category>> =
+            withContext(ioDispatcher) {
+                dataSource.getCategories()
+            }
+
+        override suspend fun getTags(categoryId: Long): Flow<List<Tag>> =
+            withContext(ioDispatcher) {
+                dataSource.getTags(categoryId)
+            }
```
`IO Dispatcher`에서 사용하기 위해 withContext를 쓰신 것 같습니다!
이 부분은 `dataSource.getCategories().flowOn(ioDispatcher)` 꼴로 사용하면 의도하신대로 동작할 뿐 아니라, withContext를 사용하지 않아 repository가 불필요한 suspend 함수가 되는 것을 막을 수 있습니다
반복이므로 기타 Repository에서는 언급하지 않겠습니다!

todayis-sunny replied:
`dataSource.getCategories().flowOn(ioDispatcher)` 오호 이런것도 있군요 배워갑니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/module/di/usecase/CategoryUseCaseModule.kt:
```diff
@@ -0,0 +1,22 @@
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object CategoryUseCaseModule {
+    @Provides
+    @Singleton
+    fun provideGetCategoriesUseCase(categoryRepository: CategoryRepository): GetCategoriesUseCase = GetCategoriesUseCase(categoryRepository)
+
+    @Provides
+    @Singleton
+    fun provideGetTagsUseCase(categoryRepository: CategoryRepository): GetTagsUseCase = GetTagsUseCase(categoryRepository)
+}
```
전에 제가 Inject 관련 의존성이 확실하지 않아서 Module로 빼야 하는 것 아니냐는 얘기를 드렸었었는데,
죄송스럽게도 `@Inject` 자체는 Android 의존성이 없어서 그냥 Usecase 자체에서 생성자 바인딩을 해도 괜찮습니다
그래서 Module 필요 없이 만드신 UseCase 자체에서 생성자 바인딩 하시면 될 것 같습니다

다른 Module 또한 마찬가지라 생략하겠습니다

todayis-sunny replied:
아하 그렇군요, 넵, 이부분도 맞게 수정해보겠습니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/navigation/BottomNavigationBar.kt:
```diff
@@ -149,6 +161,7 @@ fun BottomNavigationBar(
+@NeeGongNaeGongPreviews
 @Composable
 fun PreviewBottomNavigationBar() {
     NeeGongNaeGongTheme {
```
Neegong~Preview 달았으면 Preview 달 필요 없어보이는데 이유가 있을까용?

todayis-sunny replied:
맞습니다 기존에는 BottomNavigationBar가 높이를 지정해주지 않으면 이상하게 보여서 잘보이는 용도로 프리뷰 하나와 다크모드 라이트 모드로 각각 한개씩 총 3개를 보이게 했었는데,
내비게이션을 리팩토링하면서 이부분을 제거했었는데, 예전 커밋기록일것 같습니다!

```kotlin
@NeeGongNaeGongPreviews
@Composable
fun PreviewBottomNavigationBar() {
    NeeGongNaeGongTheme {
        BottomNavigationBar(
            navController = rememberNavController(),
        )
    }
}
```
현재 제 최종 브랜치에서는 이렇게 확인이 됩니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/create/StudiesCreateViewModel.kt:
```diff
@@ -159,6 +171,30 @@ class StudiesManagementViewModel
+            if (uiState.value.selectedTags.isEmpty()) {
+                setState { copy(validateCreateStudies = false) }
+                return
+            }
```
if ~ else if ~~
로 처리해도 되지 않나요??

todayis-sunny replied:
디버깅 해본다고 이렇게 처리 했었는데요,
```kotlin
private fun validateCreateStudies() {
    if (uiState.value.studyInfo.name.isBlank() ||
        uiState.value.studyInfo.description.isBlank() ||
        uiState.value.selectedCategory == null ||
        uiState.value.selectedTags.isEmpty()) {
        setState { copy(validateCreateStudies = false) }
        return
    }

    // 모든 유효성 검사를 통과한 경우
    setState { copy(validateCreateStudies = true) }
}
```
이런 방식이 더 나을까요 혹시?

Na2te replied:
Return 보다 else 문으로 명시하는게 더 가독성이 좋을 것 같습니다!

todayis-sunny replied:
```kotlin
private fun validateCreateStudies() {
            if (uiState.value.studyInfo.name
                    .isBlank() ||
                uiState.value.studyInfo.description
                    .isBlank() ||
                uiState.value.selectedCategory == null ||
                uiState.value.selectedTags.isEmpty()
            ) {
                setState { copy(validateCreateStudies = false) }
            } else {
                setState { copy(validateCreateStudies = true) }
            }
        }
```
이 방식으로 작성하겠습니다!

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/edit/StudiesEditViewModel.kt:
```diff
@@ -0,0 +1,233 @@
+                            presignedUrl = uiState.value.presignedUrl!!,
+                            request = uiState.value.requestImage!!,
+                        )
+                    }
```
이렇게 Null 체크하면 compiler가 uiState.value.presignedUrl하고 image를 알아서 Non Nullable한 값으로 스마트 캐스팅할텐데 nullable한 값으로 인지되고 있나요?? !!로 명시할 필요가 없지 않나요??

todayis-sunny replied:
저도 이게 의문이었습니다.
<img width="1085" alt="image" src="https://github.com/user-attachments/assets/9a763be5-5cea-4849-bbc7-7f1f7f5477ae" />
AI한테 물어보니
<img width="762" alt="image" src="https://github.com/user-attachments/assets/d410f9ce-d414-4399-a4aa-6df025971fd3" />
라는 답변을 하는군요
uiState값을 새로운 상태로 바꿔버렸을 가능성 때문에 저렇게 해버렸더라구여
컴파일러가 정말 똑똑한거 같습니다

Na2te replied:
혹시 Contract 내의 State가 Unstable한 것 같은데 @Stable 같은거로 State를 stable하게 해도 똑같나요? 신기하네용

todayis-sunny replied:
> 혹시 Contract 내의 State가 Unstable한 것 같은데 `@stable` 같은거로 State를 stable하게 해도 똑같나요? 신기하네용

State에 `@stable`을 명시하는 방식일까요?

Na2te replied:
네네 @Stable을 명시하면 됩니다
Compose runtime 관련해서 있을거에용

Na2te commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/role/StudiesMemberRoleScreen.kt:
```diff
@@ -0,0 +1,207 @@
+            onDismiss = onExpelMemberDialogDismiss,
+            onConfirm = onExpelMemberDialogConfirm,
+        )
+    }
```
userId, profileImage경로 등 많은 것들이 지금 nullable할 때 대체 Value를 넣는 식으로 가는 것 같은데,
조건문으로 null이 아닌 걸 체크하게 되면 내부에서 NonNullable한 값으로 스마트 캐스팅이 되는 걸로 알고 있는데,
차라리 NonNullable할 때 Dialog에 인자를 던져주고 보여주는 쪽으로 가는게 맞지 않았을까 싶은 생각이 있습니당

todayis-sunny replied:
저도 그런 방향쪽으로 생각해보는게 좋은거 같습니다!
그렇다면 if문 안에서 non-null체크를 검사하고 Dialog를 띄우면 되는걸까요?

Na2te replied:
Nullable한 값이 들어오는게 정상 작동 범주가 아니라면 그게 맞지 않을까 싶습니다!

upsk1 reviewed (approved):
그동안 작업 굉장히 많아보이고 한번에 다 엮어보여서 쉽지 않으셨을텐데 고생많이 하셨습니다

Aloe-droid reviewed (approved):
고생많으셨습니다 👏 👏 👏

todayis-sunny committed:
refactor(#60): StudiesApi 접두사 처리

- "/api/study-group" 공통 부분을 PREFIX로 변환

todayis-sunny committed:
refactor(#60): BaseViewModel의 handleException 네이밍 변경

- defaultHandleException로 변경
- 상속 handleException와 의미 혼동 방지

todayis-sunny committed:
refactor(#60): StudiesCreateViewModel 유효성 검사 if else 처리

todayis-sunny committed:
refactor(#60): String 긴 거를 쪼개서 합치기 (detekt 위반)

todayis-sunny committed:
refactor(#60): StudiesDrawer log 제거

todayis-sunny committed:
refactor(#60): 다이어로그 NonNullable 하게 상태 처리
거

todayis-sunny committed:
refactor(#60): issuePresignedUrl 메서드 nullable 체크

todayis-sunny committed:
refactor(#60): StudiesDrawer 부분 navBackStackEntry, isStudiesDetail 예외 return으로 반환

- navBackStackEntry는 LaunchedEffect 때문에 스터디 화면 진입시 깜빡이는 현상 발생

todayis-sunny committed:
refactor(#60): StudiesDrawer 깜빡임 현상 해결 (기존 복구)
